### PR TITLE
Add extensive test of TClass init for std::pair

### PR DESCRIPTION
**Pull request by pcanal**

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build just on windows10/cxx14, ROOT-debian10-i386/cxx14

**Comment by pcanal** [comment]
[](https://github.com/root-project/roottest/pull/869#issuecomment-1101345753)@phsft-bot build just on windows10/cxx14, ROOT-debian10-i386/cxx14

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
@phsft-bot build just on ROOT-performance-centos8-multicore/default

File: root/io/stdpair/PairLinkDef.h [hunk 1]
```diff
@@ -0,0 +1 @@
+#pragma link C++ class std::pair<reco::Muon::MuonTrackType,edm::Ref<vector<reco::Track>,reco::Track,edm::refhelper::FindUsingAdvance<vector<reco::Track>,reco::Track> > >+;
```
Comment (jblomer) [inline root/io/stdpair/PairLinkDef.h:1]: There is a missing newline

File: root/io/stdpair/cmspair.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
+//
+// Odd pair:
+//   pair<reco::Muon::MuonTrackType,edm::Ref<vector<reco::Track>,reco::Track,edm::refhelper::FindUsingAdvance<vector<reco::Track>,reco::Track> > >
+// from map:
+//   map<reco::Muon::MuonTrackType,edm::Ref<vector<reco::Track>,reco::Track,edm::refhelper::FindUsingAdvance<vector<reco::Track>,reco::Track> > >"
+
+#if 0
```
Comment (jblomer) [inline root/io/stdpair/cmspair.h:7]: I assume this block is supposed to document something but it's not clear to me what it is.

File: root/io/stdpair/pairEnumEvo.C [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      TClass *mcl = TClass::GetClass(gNameOfMap);
+      fprintf(stdout, "The map class has state: %d and classinfo %d\n", mcl->GetState(), mcl->HasInterpreterInfo());
+
+      pcl0->GetStreamerInfos()->ls();
+   }
+}
+
+void printElements(TVirtualStreamerInfo *info)
+{
+   //fprintf(stdout, "Printing StreamerInfo %p\n", info);
```
Comment (jblomer) [inline root/io/stdpair/pairEnumEvo.C:40]: Remove?

File: root/io/stdpair/pairEnumEvo.C [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+   if (libtype == libEnum::kACLiC)
+      gROOT->ProcessLine(".L cmspair.h+");
+   else if (libtype == libEnum::kShared)
+      gSystem->Load("libCmsPairCollection.so");
+   else if (libtype == libEnum::kPairShared)
+      gSystem->Load("libCmsPair.so");
+   else if (libtype == libEnum::kInterpreted)
+      gROOT->ProcessLine("#include \"cmspair.h\"");
+   else if (libtype == libEnum::kNothing)
+      {} // Do notthing.
```
Comment (jblomer) [inline root/io/stdpair/pairEnumEvo.C:87]:
```suggestion
      {} // Do nothing.
```

File: root/io/stdpair/pairEnumEvo.C [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+   if (readbeforeload) {
+      TFile *f1 = nullptr;
+      TFile *f2 = nullptr;
+      if (reportedorder)
+         f1 = TFile::Open("aa02.root");
+      f2 = TFile::Open("ab02.root");
+      if (!reportedorder)
+         f1 = TFile::Open("aa02.root");
+      // printInfo("After the early file opening.");
```
Comment (jblomer) [inline root/io/stdpair/pairEnumEvo.C:70]: Remove?

File: root/io/stdpair/pairEnumEvo.C [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+      TIter    next(info->GetElements());
+      while (TNamed *obj = (TNamed*)next())
+         if (0 != strcmp("Emulation", obj->GetTitle()))
+            obj->SetTitle("");
+   }
+
+   info->ls();
+}
+
+int pairEnumEvo(int libtype /* 0 shared, 1 ACLiC, 2 interpreted */, bool fixed, bool readbeforeload, bool reportedorder)
```
Comment (jblomer) [inline root/io/stdpair/pairEnumEvo.C:51]:
```suggestion
int pairEnumEvo(int libtype /* used as enum libEnum */, bool fixed, bool readbeforeload, bool reportedorder)
```

File: root/io/stdpair/pairEnumEvo.C [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+So at that point, it might actually be better to recreate the TClass for the
+map ...
+
+But wait ... there is currently no support for generating a collection proxy for
+an interpreted class ... so it is actually an emulated collection proxy ...
+
+That proxy does not match the interpreted (nor the compiled) version of the
+map ... so there is no good point to match the pair either ....
+
+So the solution above are (a tad bit) complex and .... not enough ...
```
Comment (jblomer) [inline root/io/stdpair/pairEnumEvo.C:175]: I'm not sure about the conclusion. This is a "won't fix"? Or the test has to wait until the issue is addressed in ROOT?

**Review by jblomer — CHANGES_REQUESTED** [review]
The commented debug `printf`s may not be valuable anymore once the test succeeds.

File: root/io/stdpair/cmspair.h (hunk 2 above)
Comment (pcanal) [reply root/io/stdpair/cmspair.h:7]: Yes, it is the distillation of the original problem (in full CMSSW) in a simple script.  I extended the comments.

File: root/io/stdpair/pairEnumEvo.C (hunk 7 above)
Comment (pcanal) [reply root/io/stdpair/pairEnumEvo.C:164]: I expanded the note to clarify that the behavior is okay for now but will need to be improved (and thus the test updated) when we support proper CollectionProxy for interpreted STL.

**Comment by pcanal** [comment]
The test is currently failing because (I think) I am missing the add_custom_command part of:
```
if(MSVC)
    ROOT_LINKER_LIBRARY(base base_cling.cxx testobject.cpp testobjectderived.cpp TEST LIBRARIES libCore libRIO)
    add_custom_command(TARGET base POST_BUILD
        COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/libbase.dll
                                         ${CMAKE_CURRENT_BINARY_DIR}/libbase.dll
        COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/libbase.lib
                                         ${CMAKE_CURRENT_BINARY_DIR}/libbase.lib
```

@bellenot Wouldn't it make sense to fold this into `ROOT_LINKER_LIBRARY` ?

**Comment by bellenot** [comment]
@pcanal probably not in `ROOT_LINKER_LIBRARY`. I'll have to make some tests to see if there is a global solution without breaking the users code

**Comment by pcanal** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by pcanal** [comment]
@phsft-bot build just on windows10/cxx14

**Pull request merged**